File: 2016/Day3/main.cpp
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

bool is_valid_triangle(int a, int b, int c) {
    return (a + b > c) && (a + c > b) && (b + c > a);
}

std::vector<std::string> read_file(const std::string& filename) {
    std::ifstream input_file(filename);
    if (!input_file) {
        std::cerr << "Unable to open file" << std::endl;
        exit(1);
    }

    std::string line;
    std::vector<std::string> lines;

    while (std::getline(input_file, line)) {
        lines.push_back(line);
    }
    input_file.close();

    return lines;
}

int count_valid_triagles_row(const std::vector<std::vector<int>>& triangles) {
    int valid_triangles = 0;
    for (const auto& triangle : triangles) {
        if (is_valid_triangle(triangle[0], triangle[1], triangle[2])) {
            valid_triangles++;
        }
    }

    return valid_triangles;
}

int count_valid_triangles_column(const std::vector<std::vector<int>>& triangles) {
    int valid_triangles = 0;
    for (size_t i = 0; i < triangles.size(); i += 3) {
        if (i + 2 < triangles.size()) {
            for (size_t j = 0; j < 3; j++) {
                int a = triangles[i][j];
                int b = triangles[i + 1][j];
                int c = triangles[i + 2][j];
                if (is_valid_triangle(a, b, c)) {
                    valid_triangles++;
                }
            }
        }
    }

    return valid_triangles;
}

int main() {
    auto lines = read_file("input.txt");

    std::vector<std::vector<int>> triangles;
    for (const auto& line : lines) {
        std::istringstream iss(line);
        std::vector<int> sides(3);
        iss >> sides[0] >> sides[1] >> sides[2];
        triangles.push_back(sides);
    }

    std::cout << "Part 1: " << count_valid_triagles_row(triangles) << std::endl;
    std::cout << "Part 2: " << count_valid_triangles_column(triangles) << std::endl;

    return 0;
}

File: 2016/Day3/main.py
import sys


def is_valid_triangle(a: int, b: int, c: int) -> bool:
    return (a + b > c) and (a + c > b) and (b + c > a)

def parse_input(lines: list[str]) -> list[list[int]]:
    triangles: list[list[int]] = []
    for line in lines:
        parts = line.split()
        if len(parts) == 3:
            sides = [int(parts[0]), int(parts[1]), int(parts[2])]
            triangles.append(sides)
    return triangles

def count_valid_triangles_row(data: list[list[int]]) -> int:
    count = 0
    for sides in data:
        if is_valid_triangle(sides[0], sides[1], sides[2]):
            count += 1
    return count

def count_valid_triangles_column(data: list[list[int]]) -> int:
    count = 0
    num_rows = len(data)

    if num_rows == 0:
        return 0

    for col_idx in range(3):
        for row_start_idx in range(0, num_rows, 3):
            if row_start_idx + 2 < num_rows:
                a = data[row_start_idx][col_idx]
                b = data[row_start_idx + 1][col_idx]
                c = data[row_start_idx + 2][col_idx]
                if is_valid_triangle(a, b, c):
                    count += 1

    return count

if __name__ == '__main__':
    input_lines = [line.strip() for line in sys.stdin if line.strip()]
    if not input_lines:
        raise ValueError("No input provided")

    data = parse_input(input_lines)
    if not data:
        raise ValueError("No valid triangle data found")

    print(f'Part 1: {count_valid_triangles_row(data)}')
    print(f'Part 1: {count_valid_triangles_column(data)}')
